### PR TITLE
[stdlib] Add a version number for the stdlib itself; use it in unit tests

### PR DESCRIPTION
**Pull request by lorentey**
The usual OS-based availability checks aren't always appropriate -- the stdlib we're actually using may not correspond to the version that shipped with the OS we're running on. This makes it tricky to decide whether a particular fix is supposed to be present in the current stdlib, which is a problem for unit tests that are supposed to test these changes.

Add an (arbitrary) version number for the stdlib itself, and use it to skip certain tests that are checking for behavioral changes introduced in 5.1.

rdar://problem/50150948

### CODE REVIEW

File: stdlib/public/core/Availability.swift [hunk 1]
```diff
@@ -45,3 +45,39 @@ public func _stdlib_isOSVersionAtLeast(
...
 #endif
 }
+
+
+/// Returns the version number for the Swift Standard Library that the code
+/// calling this was originally compiled with.
+///
+/// The version number is an arbitrary `Double` value that is monotonically
+/// increasing between toolchain releases. It does not correlate with toolchain
+/// or OS version numbers, and it is not a semantic version number.
```
Comment (milseman) [inline stdlib/public/core/Availability.swift]: If it's arbitrary and opaque and has no real correlation, why a Double?

File: stdlib/private/StdlibUnittest/StdlibVersion.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+@_fixed_layout
```
Comment (milseman) [inline stdlib/private/StdlibUnittest/StdlibVersion.swift]: If this is just for testing, why `@_fixed_layout`? Why lock in anything?

File: stdlib/public/core/Availability.swift [hunk 3]
```diff
@@ -45,3 +45,39 @@ public func _stdlib_isOSVersionAtLeast(
...
+@_alwaysEmitIntoClient @_transparent
```
Comment (milseman) [inline stdlib/public/core/Availability.swift]: Why transparent? If we always build StdlibUnittest with a just-built stdlib, why AEIC?

File: stdlib/public/core/Availability.swift [hunk 4]
```diff
@@ -45,3 +45,39 @@ public func _stdlib_isOSVersionAtLeast(
...
+public var _stdlibStaticVersion: Double {
+  return 1001.0
```
Comment (milseman) [inline stdlib/public/core/Availability.swift]: What is the maintenance story here? When does this increment? How do we track and make sure it's always incremented?

Or is it only incremented when needed for testing? In that case it is not monotonically incrementing between toolchain releases, as a single digit can span multiple toolchain releases.

File: stdlib/public/core/Availability.swift (hunk 1 above)
Comment (lorentey) [reply stdlib/public/core/Availability.swift]: It's not really opaque; it's intended as a way to create two-level version number.

The fractional part leaves some room to fork the version number progression when we open a new release branch.

```
master:                   3727.0, 3728.0, 3729.0, ...
swift-5.1-branch:         3728.0, 3728.1, 3728.2, ..., 3728.9, 3728.91, ...    
```

Of course, a pair of integers would do this a lot cleaner. I'll change it.

File: stdlib/private/StdlibUnittest/StdlibVersion.swift (hunk 2 above)
Comment (lorentey) [reply stdlib/private/StdlibUnittest/StdlibVersion.swift]: It doesn't really matter; I'll remove it.

File: stdlib/public/core/Availability.swift (hunk 4 above)
Comment (lorentey) [reply stdlib/public/core/Availability.swift]: We would increment it whenever we feel like it. Ideally we'd bump the version whenever we change stdlib behavior in an observable way. In practice, I expect we'd often leave it unchanged until we run into an issue.

It's fine to leave the version number unchanged between releases. Monotonically increasing functions are only required not to decrease.

File: stdlib/public/core/Availability.swift (hunk 3 above)
Comment (lorentey) [reply stdlib/public/core/Availability.swift]: `@_transparent` is unnecessary, but removing `@_alwaysEmitIntoClient` would defeat the purpose of this symbol -- we want the actual number to be compiled into client binaries.

I added this to simplify detection of forward/backward deployment scenarios without baking the "current" version number into tests.

**Comment by lorentey** [comment]
@swift-ci please test

File: stdlib/public/core/Availability.swift (hunk 4 above)
Comment (lorentey) [reply stdlib/public/core/Availability.swift]: I updated the comments to clarify semantics a little more.

**Comment by lorentey** [comment]
These version numbers would form a progression like this:

```
master:   (1024, 0) → (1025, 0) → (1026, 0) → ...
swift-<version>-branch:     (1025, 0) → (1025, 1) → ...
```

We would need to bump the version whenever we introduce a behavioral change that we want to be able to reliably test.

In this setup, versions corresponding to actual toolchain releases would be whatever the version was on the tag the toolchain was built from: say, `(1025, 24)` or `(1329, 128)`. 

These values would explicitly **not** correspond to major/minor/patch version numbers in some semantic versioning scheme. 

Points to ponder:

* Instead of such versioning, would it be enough to know if the current stdlib was loaded from /usr/lib/swift or a custom toolchain build?
    - I suspect it would, but it could be challenging to ensure we get the logic right every time.

* How often will this cause merge conflicts? 
    - Cherry picking to release branches will be more annoying, and PRs will conflict more often.
    - We should make sure routine version bumps never lead to automerger failures -- the version must be kept in sync between branches that use change propagation.

* Do we keep tests indefinitely gated at random `.custom(2312, 13)` version numbers, or do we go back and update them to the released version once that's done?
    - How do we maintain the mapping between official releases and these versions?

* Should we rather implement a proper semantic versioning scheme?

* Is there a way to tie the availability of stdlib's own features to these finer-grained versions, rather than magical 9999 OS versions?

**Comment by lorentey** [comment]
@swift-ci please test

File: stdlib/public/core/Availability.swift [hunk 5]
```diff
@@ -45,3 +45,57 @@ public func _stdlib_isOSVersionAtLeast(
   return false._value
 #endif
 }
+
+@inlinable
```
Comment (milseman) [inline stdlib/public/core/Availability.swift]: Why inlinable?

File: stdlib/public/core/Availability.swift [hunk 6]
```diff
@@ -45,3 +45,57 @@ public func _stdlib_isOSVersionAtLeast(
...
+/// The version number is an arbitrary pair of integers, with lexicographical
+/// ordering. Version numbers of (logically) successive stdlib releases form a
+/// monotonically increasing sequence; i.e., versions should not decrease, but
+/// they are allowed to stay the same.
+///
+/// The two integer components are not intended to correspond to major or minor
+/// versions in a semantic versioning scheme. Neither do they correlate with
+/// toolchain or OS version numbers.
+public var _stdlibStaticVersion: (Int, Int) {
+  @_alwaysEmitIntoClient
```
Comment (milseman) [inline stdlib/public/core/Availability.swift]: Why do we want back deployment?

File: stdlib/public/core/Availability.swift [hunk 7]
```diff
@@ -45,3 +45,57 @@ public func _stdlib_isOSVersionAtLeast(
...
+///
+/// The version number is an arbitrary pair of integers, with lexicographical
+/// ordering. Version numbers of (logically) successive stdlib releases form a
+/// monotonically increasing sequence; i.e., versions should not decrease, but
+/// they are allowed to stay the same.
+///
+/// The two integer components are not intended to correspond to major or minor
+/// versions in a semantic versioning scheme. Neither do they correlate with
+/// toolchain or OS version numbers.
+@_alwaysEmitIntoClient // Introduced in 5.1
```
Comment (milseman) [inline stdlib/public/core/Availability.swift]: Why do we want back-deployment?

File: stdlib/public/core/Availability.swift [hunk 8]
```diff
@@ -45,3 +45,57 @@ public func _stdlib_isOSVersionAtLeast(
...
+    return _stdlibOpaqueVersion
+  }
+  else {
+    // When linked with the 5.0 stdlib, we return this default value.
+    return (1000, 0)
+  }
+}
+
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)
+@usableFromInline
```
Comment (milseman) [inline stdlib/public/core/Availability.swift]: Not needed if not back deploying

**Review by milseman — CHANGES_REQUESTED** [review]
I still don't see why we need to back-deploy or bake any implementation details into our ABI.

File: stdlib/public/core/Availability.swift (hunk 6 above)
Comment (lorentey) [reply stdlib/public/core/Availability.swift]: In this case, `@_alwaysEmitIntoClient` is not really here to allow back-deployment -- it's to get rid of the symbol in libswiftCore. This is akin to a header-only `#define` constant; the static version number should never ever be linked in from the currently running stdlib.

File: stdlib/public/core/Availability.swift (hunk 7 above)
Comment (lorentey) [reply stdlib/public/core/Availability.swift]: Is there a reason we shouldn't?

File: stdlib/public/core/Availability.swift (hunk 6 above)
Comment (milseman) [reply stdlib/public/core/Availability.swift]: AEIC gets rid of the symbol for the getter. You're just swapping it for `_stdlibOpaqueVersion` in exchange for more stuff in our binary compatibility story. How is that better?

Comment (lorentey) [reply stdlib/public/core/Availability.swift]: I still don't get it -- `_stdlibStaticVersion` is a compile-time constant for the version of the stdlib that the code was compiled with. By definition, it is public API that must not ever become an ABI symbol in libswiftCore. `@_alwaysEmitIntoClient` is perfect for it.

`_stdlibOpaqueVersion` is different -- it returns the version of the stdlib that is loaded at runtime, so it must be non-inlinable, with an availability.

Comment (lorentey) [reply stdlib/public/core/Availability.swift]: Oh wait, I see what you mean -- we're swapping it for `_stdlibCurrentVersion`, not `_stdlibOpaqueVersion`. Yeah, that needs to be fixed; I apparently forgot to get rid of it.

File: stdlib/public/core/Availability.swift [hunk 9]
```diff
@@ -45,3 +45,52 @@ public func _stdlib_isOSVersionAtLeast(
...
+///
+/// The version number is an arbitrary pair of integers, with lexicographical
+/// ordering. Version numbers of (logically) successive stdlib releases form a
+/// monotonically increasing sequence; i.e., versions should not decrease, but
+/// they are allowed to stay the same.
+///
+/// The two integer components are not intended to correspond to major or minor
+/// versions in a semantic versioning scheme. Neither do they correlate with
+/// toolchain or OS version numbers.
+public var _stdlibStaticVersion: (Int, Int) {
```
Comment (milseman) [inline stdlib/public/core/Availability.swift]: Still a little confused. Why is this public?

Comment (lorentey) [reply stdlib/public/core/Availability.swift]: It allows tests to easily determine if they are back/forward deployed. Binaries that are intimately tied to a particular revision of the stdlib and should not be separately back/forward deployed (e.g. libswiftFoundation) could use it for sanity checks.

This is not strictly necessary, but it seems like a cheap addition.

Comment (lorentey) [reply stdlib/public/core/Availability.swift]: FWIW, dyld provides two similar version numbers through `NSVersionOfRunTimeLibrary` and `NSVersionOfLinkTimeLibrary`. 

(The link-time version queries the linkage of the main executable, not the current object, so it's different(ly useful) than the static version number implemented here.)

**Comment by milseman** [comment]
@lorentey and I discussed this. If this is just used for our own testing, we can keep it internal so long as StdlibUnittest is built with disable access control.

Karoy, what do you think about making this proper API at some point?

**Comment by lorentey** [comment]
Closing; adding 9999 availability guards to such behavioral tests already covers the most important deployment scenarios. (#24496) A version number like this would be helpful, but its maintenance cost would exceed its usefulness in its current form. (E.g., this variant does not plug into `@available` syntax.)

**Comment by lorentey** [comment]
@milseman I think all resilient Swift modules should have their own separate version number, and there should be a standard way to access it (both at compile-time and at runtime). The version number should be usable in compile-time `@available` declarations and runtime `#available` checks. Also, the version number should match the one used by SPM, which implies a semantic versioning scheme.

**Pull request closed without merging**